### PR TITLE
Added custom colors and sorting by color and ddg

### LINKED ISSUE #120: ddg2logo revision 2
once we solve the issue "ddg2logo revision 1", we could include a config file for defining residues in classes of interest
and use the logo with these classes. For example, Ar (aromatic), Hp (Hydrophobic), etc

### PR DESCRIPTION
**Pull request by mattu90**
Fix #120

### CODE REVIEW

File: bin/ddg2logo [hunk 1]
```diff
@@ -107,6 +122,76 @@ def plot_matrix(mat, reslist, letters, xlabels, max_range, title=None, fontsize=
         # AA mutations sorted by energy
         s_pos_idx = pos_idx[0][local_pos_idx]
 
+        
+
+        #if the config file is provided, load the config 
+        if config_file is not None:
+            with open(options.config_file) as f:
+                config = yaml.load(f, Loader=SafeLoader)
+
```
Comment (mtiberti) [inline bin/ddg2logo:131]: I'm not reviewing this part for now - see general comment

File: bin/ddg2logo [hunk 2]
```diff
@@ -191,7 +276,7 @@ optional.add_argument("-L","--png-letter-map",dest='pngfile',action='store', def
 thresholds.add_argument("-T","--stab-threshold", dest="threshold_stab", action='store', type=float, help="DDG threshold values for stabilizing mutations - mutations with DDG higher than this are not considered.")
 thresholds.add_argument("-t","--destab-threshold", dest="threshold_destab", action='store', type=float, help="DDG threshold values for destabilizing mutations - mutations with DDG lower than this are not considered.")
 optional.add_argument("-x","--maximum-degeneracy", dest="degeneracy", action='store', type=int, default=20, help="when the number of stabilizing/destabilizing mutations is higher than this threshold an X will be plotted instead")
-
+optional.add_argument("-c","--config", dest="config_file", action='store', help="configuration file with which the user can set the colour of subsets of residues")
```
[the remark this answers is not in the document]
Comment (mtiberti) [reply bin/ddg2logo:279]: In particular, we should be sure that if no config file is specified the colors are used as-is (or we can define a custom default mapping to be applied, e.g. all black)

**Review by mtiberti — CHANGES_REQUESTED** [review]
Hi, thank you - I think we need to change the config file format since right now it's really difficult to understand for the uninitiated. I think we could try something like this:

```
hydrophobic:
    residues:
        - W
        - F
    color: 0.1 0.3 0.0

negatively_charged:
    residues:
        - E
        - D
    color: 0.0 0.5 0.7
```

which parses to:

```
{'hydrophobic': {'residues': ['W', 'F'], 'color': '0.1 0.3 0.0'},
 'negatively_charged': {'residues': ['E', 'D'], 'color': '0.0 0.5 0.7'}}
```

I think this is handy to maintain the color-residue type association and shouldn't be too difficult to integrate in the current code.

File: bin/ddg2logo [hunk 3]
```diff
@@ -34,6 +34,8 @@ from mutatex.letters import letters
 import numpy as np
 import logging as log
 import os
+import yaml
```
[the remark this answers is not in the document]
Comment (mattu90) [reply bin/ddg2logo:37]: Ok, I added it

File: bin/ddg2logo [hunk 4]
```diff
@@ -224,6 +309,7 @@ except:
 # parse mutation list
 try:
     res_order = parse_mutlist_file(options.mutation_list)
+    print(res_order)
```
[the remark this answers is not in the document]
Comment (mattu90) [reply bin/ddg2logo:312]: I removed it

File: bin/ddg2logo (hunk 2 above)
[the remark this answers is not in the document]
Comment (mattu90) [reply bin/ddg2logo:279]: I added it

[the remark this answers is not in the document]
Comment (mattu90) [reply bin/ddg2logo:279]: at the moment if no config is provided, the Colors are used as-is

File: bin/ddg2logo [hunk 5]
```diff
@@ -78,7 +80,20 @@ def parse_letters_png(pngfile, letters="ACDEFGHIKLMNPQRSTVWXYypshzkmloef"):
 
     return out
 
-def plot_matrix(mat, reslist, letters, xlabels, max_range, title=None, fontsize=8, hgap=0.5):
+# function to change the color of a specific residue
+def color_letter(letter, color, out): 
+
+    color = color.split(" ")
+    out = out[letter]
+    for i in range(len(out[:,:,0])):
```
[the remark this answers is not in the document]
Comment (mattu90) [reply bin/ddg2logo]: Now it's changed

File: bin/ddg2logo [hunk 6]
```diff
@@ -107,6 +122,76 @@ def plot_matrix(mat, reslist, letters, xlabels, max_range, title=None, fontsize=
         # AA mutations sorted by energy
         s_pos_idx = pos_idx[0][local_pos_idx]
 
+        
+
+        #if the config file is provided, load the config 
+        if config_file is not None:
```
[the remark this answers is not in the document]
Comment (mattu90) [reply bin/ddg2logo:125]: Added the config only in the main script and added the try/except

File: bin/ddg2logo [hunk 7]
```diff
@@ -273,6 +359,15 @@ if options.labels:
 else:
     labels = res_id_labels
 
+# use custom colour for residues subsets if a config file is provided
+if options.config_file:
```
[the remark this answers is not in the document]
Comment (mattu90) [reply bin/ddg2logo:354]: Added the config only in the main script and added the try/except

File: bin/ddg2logo [hunk 8]
```diff
@@ -319,6 +414,7 @@ for i in range(len(spliced_data)):
                         max_range,
                         title=options.title,
                         fontsize=options.fontsize,
+                        config_file=options.config_file
                     )
     plt.show()
```
[the remark this answers is not in the document]
Comment (mattu90) [reply bin/ddg2logo:422]: Maybe I don't understand, but I thing it is correct

File: bin/ddg2logo [hunk 9]
```diff
@@ -78,7 +80,20 @@ def parse_letters_png(pngfile, letters="ACDEFGHIKLMNPQRSTVWXYypshzkmloef"):
 
     return out
 
-def plot_matrix(mat, reslist, letters, xlabels, max_range, title=None, fontsize=8, hgap=0.5):
+# function to change the color of a specific residue
+def color_letter(letter, color, out):
```
[the remark this answers is not in the document]
Comment (mattu90) [reply bin/ddg2logo]: I changed the name but also, in the end, we removed the out from color_letter

File: bin/ddg2logo [hunk 10]
```diff
@@ -273,6 +359,15 @@ if options.labels:
...
     labels = res_id_labels
 
+# use custom colour for residues subsets if a config file is provided
+if options.config_file:
+    with open(options.config_file) as f:
+        config = yaml.load(f, Loader=SafeLoader)
+    for key in config:
+        for color, residues in key.items():
+            for residue in residues:
+                color_letter(residue, color, letters)
```
[the remark this answers is not in the document]
Comment (mattu90) [reply bin/ddg2logo]: Changed in the script

File: bin/ddg2logo (hunk 1 above)
Comment (mattu90) [reply bin/ddg2logo]: ok

File: bin/ddg2logo (hunk 8 above)
[the remark this answers is not in the document]
Comment (mtiberti) [reply bin/ddg2logo:422]: let's remove line 422 here - it's not necessary (my bad)

File: bin/ddg2logo [hunk 11]
```diff
@@ -78,7 +80,17 @@ def parse_letters_png(pngfile, letters="ACDEFGHIKLMNPQRSTVWXYypshzkmloef"):
...
+def color_letter(letter, color): 
+    new_array = np.copy(letter)
+    
+    new_array[:,:,0][np.where(new_array[:,:,0] != 1.0)] = color[0]
+    new_array[:,:,1][np.where(new_array[:,:,1] != 1.0)] = color[1]
+    new_array[:,:,2][np.where(new_array[:,:,2] != 1.0)] = color[2]
+
+    return new_array
+
+def plot_matrix(mat, reslist, letters, xlabels, max_range, title=None, fontsize=8, config_file=None, hgap=0.5):
```
Comment (mtiberti) [inline bin/ddg2logo:93]: this should have argument `config` rather than `config_file`

File: bin/ddg2logo [hunk 12]
```diff
@@ -273,6 +350,27 @@ if options.labels:
...
+        exit(1)
+
+    new_letters = letters.copy()
+
+    for key in config:
+        for aa in config[key]['residues']:
+            new_letters[aa] = color_letter(new_letters[aa], config[key]['color'])
+
+    letters = new_letters
+
```
Comment (mtiberti) [inline bin/ddg2logo:369]: extra newlines should be removed

File: bin/ddg2logo [hunk 13]
```diff
@@ -319,6 +417,8 @@ for i in range(len(spliced_data)):
                         max_range,
                         title=options.title,
                         fontsize=options.fontsize,
+                        config_file=options.config_file
```
Comment (mtiberti) [inline bin/ddg2logo:420]: this should have `config=config` rather than the file

File: doc/manual.lyx [hunk 14]
```diff
@@ -424,7 +424,7 @@ $NRUNS$
 \end_inset
```
Comment (mtiberti) [inline doc/manual.lyx:425]: I think you need to merge `master` into branch and then reapply your changes - I see some parts are being removed with merge here

**Review by mtiberti — CHANGES_REQUESTED** [review]

**Review by mtiberti — APPROVED** [review]

**Pull request merged**